File: leetcode/19.remove-nth-node-from-end-of-list/ts/removenthfromend.ts
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

// N, 1
const removeNthFromEnd = (head: ListNode | null, n: number): ListNode | null => {
    const dummy: ListNode | null = new ListNode(0, head);

    let curr = head
    let prevSlow: ListNode | null = dummy

    let i = 0
    while (curr != null) {
        if (i++ >= n) prevSlow = prevSlow!.next
        curr = curr.next
    }

    prevSlow!.next = prevSlow!.next!.next

    return dummy.next
};

export default removeNthFromEnd

File: leetcode/19.remove-nth-node-from-end-of-list/go/removenthfromend.go
package removefromend

type ListNode struct {
	Val  int
	Next *ListNode
}

// N, 1
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{0, head}

	curr := head
	slowPrev := dummy

	i := 0

	for curr != nil {
		if i >= n {
			slowPrev = slowPrev.Next
		}
		i++
		curr = curr.Next
	}

	slowPrev.Next = slowPrev.Next.Next
	return dummy.Next
}
